File: src/auto_tagger/llm/schemas.py
"""Pydantic schemas for structured LLM responses."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CandidateSelectionResponse(BaseModel):
    """Structured response for candidate selection."""

    model_config = ConfigDict(extra="forbid")

    selected_index: int | None = None
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str

    def validate_candidate_count(self, candidate_count: int) -> None:
        """Validate selected index against number of candidates."""
        if self.selected_index is None:
            return
        if self.selected_index < 0 or self.selected_index >= candidate_count:
            raise ValueError(f"selected_index out of range for {candidate_count} candidates")


class GeneratedTrackTags(BaseModel):
    """Generated tags for one track."""

    model_config = ConfigDict(extra="forbid")

    title: str
    artist: str | None = None
    artists: list[str] = Field(default_factory=list)
    album: str | None = None
    album_artist: str | None = None
    track_number: int | None = Field(default=None, gt=0)
    disc_number: int | None = Field(default=None, gt=0)


class GenreEnrichmentResponse(BaseModel):
    """Structured response for genre enrichment via LLM.

    Returns a Discogs-style genre string (e.g. 'Electronic, Ambient, Modern Classical')
    or None when uncertain.
    """

    model_config = ConfigDict(extra="forbid")

    genre: str | None = None


class FolderExtractionResponse(BaseModel):
    """Structured response for folder name metadata extraction."""

    model_config = ConfigDict(extra="forbid")

    artist: str | None = None
    album: str | None = None
    year: str | None = None
    disc: str | None = None


class FallbackTagResponse(BaseModel):
    """Structured response for fallback tag generation."""

    model_config = ConfigDict(extra="forbid")

    artist: str
    artists: list[str] = Field(default_factory=list)
    album: str
    album_artist: str
    album_artists: list[str] = Field(default_factory=list)
    year: str | None = None
    genre: str | None = None
    tracks: list[GeneratedTrackTags]
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str

    @model_validator(mode="before")
    @classmethod
    def reject_musicbrainz_ids(cls, data: Any) -> Any:
        """Reject generated MusicBrainz fields."""
        if isinstance(data, dict):
            invented = [key for key in data if key.lower().startswith("musicbrainz")]
            if invented:
                raise ValueError("fallback output must not include MusicBrainz IDs")
        return data


class CorrectedTrack(BaseModel):
    """Complete corrected metadata for one track returned by LLM audit.

    Only populated for tracks with issues (warning/error). The code writes
    these values verbatim — no field mapping or transformation.
    """

    model_config = ConfigDict(extra="forbid")

    title: str | None = None
    artist: str | None = None
    artists: list[str] | None = None
    album: str | None = None
    album_artist: str | None = None
    year: str | None = None
    genre: str | None = None


class AuditTrackResult(BaseModel):
    """LLM audit result for a single track field."""

    model_config = ConfigDict(extra="forbid")

    index: int = Field(..., ge=0, description="Track index within the album")
    field: str = Field(..., description="Audited field name (artist, title, album, album_artist, artists, path)")
    status: str = Field(..., description="'correct' | 'warning' | 'error'")
    message: str = Field(..., description="Human-readable explanation")
    suggestion: str | None = Field(default=None, description="Suggested fix value (deprecated in favor of corrected)")
    corrected: CorrectedTrack | None = Field(default=None, description="Complete corrected metadata for problematic tracks")


class AuditResponse(BaseModel):
    """Structured response for LLM album audit."""

    model_config = ConfigDict(extra="forbid")

    tracks: list[AuditTrackResult] = Field(..., description="Per-track audit results")

File: frontend/electron/handlers/schemas.ts
/**
 * Structured output schemas for LLM responses.
 * Ported from Python auto_tagger.llm.schemas.
 */

// ── Candidate Selection ─────────────────────────────────────────────

export interface CandidateSelectionResponse {
  selectedIndex: number | null;
  confidence: number; // 0.0 – 1.0
  reason: string;
}

export function validateSelectionIndex(
  response: CandidateSelectionResponse,
  candidateCount: number,
): void {
  if (
    response.selectedIndex != null &&
    (response.selectedIndex < 0 || response.selectedIndex >= candidateCount)
  ) {
    throw new Error(
      `selectedIndex ${response.selectedIndex} out of range for ${candidateCount} candidates`,
    );
  }
}

// ── Generated Track Tags ────────────────────────────────────────────

export interface GeneratedTrackTags {
  title: string;
  artist?: string | null;
  artists?: string[];
  album?: string | null;
  albumArtist?: string | null;
  trackNumber?: number | null;
  discNumber?: number | null;
}

// ── Fallback Tag Response ───────────────────────────────────────────

export interface FallbackTagResponse {
  artist: string;
  artists: string[];
  album: string;
  albumArtist: string;
  albumArtists: string[];
  year?: string | null;
  genre?: string | null;
  tracks: GeneratedTrackTags[];
  confidence: number; // 0.0 – 1.0
  reason: string;
}

// ── Folder Extraction Response ──────────────────────────────────────

export interface FolderExtractionResponse {
  artist?: string | null;
  album?: string | null;
  year?: string | null;
  disc?: string | null;
}

// ── Genre Enrichment Response ───────────────────────────────────────

export interface GenreEnrichmentResponse {
  genre?: string | null;
}

// ── Audit Types ────────────────────────────────────────────────────

export interface AuditTrackResult {
  index: number;
  field: string;
  status: "correct" | "warning" | "error";
  message: string;
  suggestion?: string | null;
  corrected?: CorrectedTrack | null;
  source?: "deterministic" | "llm";
  confidence?: number;
  autoFixEligible?: boolean;
  autoFixed?: boolean;
}

export interface CorrectedTrack {
  title?: string | null;
  artist?: string | null;
  artists?: string[] | null;
  album?: string | null;
  albumArtist?: string | null;
  albumArtists?: string[] | null;
  year?: string | null;
  genre?: string | null;
  trackNumber?: number | null;
  trackTotal?: number | null;
  discNumber?: number | null;
  discTotal?: number | null;
}

export interface AuditResponse {
  tracks: AuditTrackResult[];
}
